_posts/2019-04-01-functional-introduction.md

```diff
@@ -0,0 +1,106 @@
+
+Co jakiś czas jest przedstawiana jakaś stara prawda, czy też adoptowane jest jakieś nowe narzędzie. Tak też nie jest już nowością, że programowanie funkcyjne zyskało na popularności i ułatwia wiele rzeczy. W Javie mamy różne funkcyjne bibliteki umożliwiające tworzenie bardziej funkcyjnego kodu jak chociażby Vavr, oraz JOOλ. W Kotlinie mamy Arrow choć sam język jest tutaj z natury funkcyjny. W tym wpisie zacznijmy od omówienia funkcjnego podejścia oraz jedenej z podstawowych struktr danych tam istniejących. 
+```
+
+### Szable w dłoń i robimy funkcyjnie!
+#### Ale zacznijmy od początku... 🛤 
+Funkcjnie, czyli mamy... funkcje. Mają one jakieś wejście/wyjście. Kluczowym konceptem jest tutaj modułowość. Możemy rozbić nasz algorytm, czy cokolwiek tam robimy na mniejsze pod-funkcje. Jak wiadomo nasze umysły mają tylko ograniczoną ilość RAMu do rozwiązywania codziennych problemów, dlatego wymyślamy koncepty, paradygmaty, standardy aby uprościć to wszystko. Małe moduły mogą być kodzone szybko i być łatwo testowane. Do tego jest możliwe ich re-używanie. Czyli mamy w sumie kilka najważniejszych zasad dobrego kodu spakowane do jednego worka z tego też powodu programowanie zorientowane na funkcje staje się poopularne. 
+
+
+### OOP vs FP 🥊
+Jak to zwykle bywa czasami jest hype na nowy język, framework, metodykę, czy cokolwiek innego. Zazwyczaj stoją ku temu powody i tymi powodami najczęsciej jest rozwiązanie jakiegoś problemu, który niektórym z nas akurat przeszkadzał. Niemniej ważne, że przy takich sprawach wsprowadzamy rozwiązanie do problemu, a nie problem do rozwiązania. Używanie FP wcale nie znaczy, że OOP jest już passé. Wręcz przeciwnie oba paradygmaty doskonale ze sobą współpracują. Dobra... Może nie do końca współpracują, ale bardziej zastępują lub uzpełniają niektóre techniki zawierające się w innych paradygmatach. 
+
+#### Ważniejsze cechy funkcjonalnego podejścia: 
+
+⚙ Higher-order functions - przekazują funkcję jako paramter do innej funkcji - istna incepcja.
+
+Na przykład funkcja `availableCustomers(Supplier<Boolean> customerAvailability)` przyjmuje funkcję jako parametr `Customer::isAvailable`.
+
+Czego wynikiem jest: `availableCustomers(Customer::isAvailable)`
+
+⚙ First-class citizens - czyli traktowanie funkcji jako wartości.
+
+Dla obrazowania przykładu zróbmy sobie funkcję w Kotlinie o wdzięcznej nazwie `adder`: 
+
+*  `fun add(a: Int, b: Int) = (a + b).toDouble()` - nasza funkcja.
+*  `val adder = ::add` - którą przypisujemy do zmiennej.
+*  `val adder: (Int, Int) -> Double = { a, b -> (a + b).toDouble() }` - lub to samo w ten sposób.
+*  `adder(1,1)` - po czym robimy skomplikowane obliczenia.
+
+⚙ Anonymous classes - `() -> "czyli lambdy"`
+
+### Skutki uboczne programowania imperatywnego 🔰
+Impratywnego, czyli takiego z jakim mamy styczność zazwyczaj na początku drogi z programowaniem.
```

Comment:
lepiej byłoby napisać co to jest programowanie imperatywne

Reply:
outdated